scan no cols

adrian-ionescu opened the pull request:
This change addresses two issues:

1) The provider was unable to handle "scan" requests for which the column list was empty.
- The problem was that the constructed SparkSql query looked something like "select  from table".
- Fixed by performing a "select count(*) from table", using the result to create an RDD of Row.empty, registering that as a temp_table_no_cols, and finally executing a "select \* from temp_table_no_cols".

2) Temp tables registered with the sqlContext were never dropped.
- The problem was that we simply never called sqlContext.dropTable()
- Fixed by implementing a TempTable class with a .close() method, so that we can treat instances of it as managed resources (in the Scala ARM sense).

and-costea commented on .gitignore:
```diff
@@ -12,5 +12,5 @@ target/
 project/project
-/loader/
-/provider/
+/project/
```
I know the previous ones didn't make sense, but neither do these.

and-costea commented on loader/src/main/scala/com/actian/spark_vector/loader/spark_sql_utils/SparkSqlTable.scala:
```diff
@@ -0,0 +1,32 @@
+package com.actian.spark_vector.loader.spark_sql_utils
```
This makes more sense to be in com.actian.spark_vector.sql

Also, in package names _ replaces -.

and-costea commented on loader/src/main/scala/com/actian/spark_vector/loader/command/ConstructVector.scala:
```diff
@@ -46,9 +47,9 @@ object ConstructVector {
-    // Load the line item data into Vector
-    sqlContext.sql(s"insert into table ${targetTempTable} ${select}")
+    for (targetTempTable <- managed(VectorTempTable.register(config, sqlContext))) {
+      // Load the line item data into Vector
```
remove

and-costea commented on loader/src/main/scala/com/actian/spark_vector/loader/command/ConstructVector.scala:
```diff
@@ -46,9 +47,9 @@ object ConstructVector {
-    sqlContext.sql(s"insert into table ${targetTempTable} ${select}")
+    for (targetTempTable <- managed(VectorTempTable.register(config, sqlContext))) {
+      // Load the line item data into Vector
+      sqlContext.sql(s"insert into table ${targetTempTable.quoted_name} ${select}")
```
convention in Scala/Java is camelCase... quotedName

and-costea commented on loader/src/main/scala/com/actian/spark_vector/loader/spark_sql_utils/package.scala:
```diff
@@ -15,8 +15,13 @@
   def sparkQuote(name: String): String = s"`$name`"
 
-  def colsSelectStatement(cols: Option[Seq[String]]): String = cols.map(_.map(sparkQuote(_)).mkString(", ")).getOrElse("*")
+  def colsSelectStatement(cols: Option[Seq[String]]): String = {
```
Also, you can get rid of one pair of {}: def selectStatement(...) = cols match {
}

and-costea commented on loader/src/main/scala/com/actian/spark_vector/loader/spark_sql_utils/package.scala:
```diff
@@ -15,8 +15,13 @@
-  def colsSelectStatement(cols: Option[Seq[String]]): String = cols.map(_.map(sparkQuote(_)).mkString(", ")).getOrElse("*")
+  def colsSelectStatement(cols: Option[Seq[String]]): String = {
+    cols match {
+      case None | Some(Seq()) => "*" /* it is legal in SparkSql to execute 'select * from zero_cols_table' */
```
Some(Nil) ?

and-costea commented on loader/src/main/scala/com/actian/spark_vector/loader/spark_sql_utils/package.scala:
```diff
@@ -15,8 +15,13 @@
+  def colsSelectStatement(cols: Option[Seq[String]]): String = {
+    cols match {
+      case None | Some(Seq()) => "*" /* it is legal in SparkSql to execute 'select * from zero_cols_table' */
+      case Some(seq) => seq.map(sparkQuote(_)).mkString(", ")
```
seq.map(sparkQuote).mkString(", ")

and-costea commented on loader/src/main/scala/com/actian/spark_vector/loader/spark_sql_utils/SparkSqlTable.scala:
```diff
@@ -0,0 +1,32 @@
+import org.apache.spark.sql.DataFrame
+
+sealed trait SparkSqlTable {
+  def table_name : String
```
tableName

and-costea commented on loader/src/main/scala/com/actian/spark_vector/loader/spark_sql_utils/SparkSqlTable.scala:
```diff
@@ -0,0 +1,32 @@
+
+sealed trait SparkSqlTable {
+  def table_name : String
+  def quoted_name = sparkQuote(table_name)
```
quotedName

and-costea commented on loader/src/main/scala/com/actian/spark_vector/loader/spark_sql_utils/SparkSqlTable.scala:
```diff
@@ -0,0 +1,32 @@
+sealed trait SparkSqlTable {
+  def table_name : String
+  def quoted_name = sparkQuote(table_name)
+  def close() = {}
```
def close(): Unit

and-costea commented on loader/src/main/scala/com/actian/spark_vector/loader/spark_sql_utils/SparkSqlTable.scala:
```diff
@@ -0,0 +1,32 @@
+  def close() = {}
+}
+
+case class HiveTable(override val table_name: String) extends SparkSqlTable 
```
better leave close as abstract in SparkSQLTable and force it to be defined in subclasses.

and-costea commented on loader/src/main/scala/com/actian/spark_vector/loader/spark_sql_utils/SparkSqlTable.scala:
```diff
@@ -0,0 +1,32 @@
+case class HiveTable(override val table_name: String) extends SparkSqlTable 
+
+class TempTable private (override val table_name: String, df: DataFrame) extends SparkSqlTable {
+  def register() : Unit = {
```
def register(): Unit = df.registerTempTable(...)

and-costea commented on provider/src/main/scala/com/actian/spark_vector/provider/RequestHandler.scala:
```diff
@@ -16,23 +16,31 @@
+import org.apache.spark.SparkContext
+import com.actian.spark_vector.loader.spark_sql_utils.TempTable
+import com.actian.spark_vector.loader.spark_sql_utils.SparkSqlTable
+import com.actian.spark_vector.loader.spark_sql_utils.HiveTable
```
do an optimize imports

and-costea commented on provider/src/main/scala/com/actian/spark_vector/provider/RequestHandler.scala:
```diff
@@ -127,45 +116,75 @@ class RequestHandler(sqlContext: SQLContext, val auth: ProviderAuth) extends Log
-    val ret = s"${prefix}_${id.incrementAndGet}"
-    df.registerTempTable(ret)
-    ret
+  private def getFormat(part: JobPart) : String = {
```
maybe make this a method to JobPart? and getFormat -> format

and-costea commented on provider/src/main/scala/com/actian/spark_vector/provider/RequestHandler.scala:
```diff
@@ -127,45 +116,75 @@ class RequestHandler(sqlContext: SQLContext, val auth: ProviderAuth) extends Log
+  }
 
-    s"select ${part.column_infos.map(ci => sparkQuote(ci.column_name)).mkString(", ")} from ${sparkQuote(table)}"
+  private def handleScan(part: JobPart) {
```
def handleScan = for { ... }

adrian-ionescu commented on provider/src/main/scala/com/actian/spark_vector/provider/RequestHandler.scala:
```diff
@@ -122,50 +103,88 @@ class RequestHandler(sqlContext: SQLContext, val auth: ProviderAuth) extends Log
-  private def writeDF(format: String, df: DataFrame, part: JobPart): Unit = {
-    val options = part.options.getOrElse(Map.empty[String, String])
+
+  /** Given a job part of type "insert" by writing into the external table all tuples coming from the VectorDF. */
```
Given -> Handle

and-costea commented on provider/src/main/scala/com/actian/spark_vector/provider/RequestHandler.scala:
```diff
@@ -122,50 +103,87 @@ class RequestHandler(sqlContext: SQLContext, val auth: ProviderAuth) extends Log
-    val ret = s"${prefix}_${id.incrementAndGet}"
-    df.registerTempTable(ret)
-    ret
+  /** Convenience function: retrieve a job part's options, if any, or else an empty Map */
```
remove convenience function (it's pretty obvious)?

and-costea commented on provider/src/main/scala/com/actian/spark_vector/provider/RequestHandler.scala:
```diff
@@ -122,50 +103,87 @@ class RequestHandler(sqlContext: SQLContext, val auth: ProviderAuth) extends Log
+   * 1) return the explicit part.format, if specified
+   * 2) return the extension of part.external_reference, if recognized
+   * 3) throw IllegalArgumentException */
+  private def getFormat(part: JobPart): String = {
```
remove one pair of {}

and-costea commented on provider/src/main/scala/com/actian/spark_vector/provider/RequestHandler.scala:
```diff
@@ -122,50 +103,87 @@ class RequestHandler(sqlContext: SQLContext, val auth: ProviderAuth) extends Log
+  /** Given a job part, return the format of the external table, according to the following logic:
+   * 1) return the explicit part.format, if specified
+   * 2) return the extension of part.external_reference, if recognized
+   * 3) throw IllegalArgumentException */
```
, otherwise

and-costea commented on provider/src/main/scala/com/actian/spark_vector/provider/RequestHandler.scala:
```diff
@@ -122,50 +103,87 @@ class RequestHandler(sqlContext: SQLContext, val auth: ProviderAuth) extends Log
-   */
-  private def selectStatement(format: String, part: JobPart): String = {
-    val options = part.options.getOrElse(Map.empty[String, String])
+  /** Given job part, return the corresponding SparkSqlTable that one may then "select * from" */
```
`SparkSqlTable`

and-costea commented on provider/src/main/scala/com/actian/spark_vector/provider/RequestHandler.scala:
```diff
@@ -122,50 +103,87 @@ class RequestHandler(sqlContext: SQLContext, val auth: ProviderAuth) extends Log
+        TempTable("src", df)
+      }
+    }
+    closeResourceOnFailure(table) {
```
if (part.column_infos.isEmpty) {
   val dfNoCols = closeResourceAfterUse(table) {
        .....
   }
   TempTable("src_no_cols", dfNoCols)
} else {
   table
}

and-costea commented on src/main/scala/com/actian/spark_vector/sql/SparkSqlTable.scala:
```diff
@@ -0,0 +1,30 @@
+package com.actian.spark_vector.sql
```
header copyright

cbarca replied:
run "sbt scalastyle" too for some extra checking

and-costea commented on src/main/scala/com/actian/spark_vector/sql/package.scala:
```diff
@@ -16,4 +16,13 @@
+package object sql {
+  def sparkQuote(name: String): String = s"`$name`"
+
+  def colsSelectStatement(cols: Option[Seq[String]]): String = {
```
= cols match {

Commit:
Implement scanning of 0 columns from external tables and dropping of all temp tables registered with the sqlContext. Fixes #55.

This change addresses two issues:

1. The provider was unable to handle "scan" requests for which the column list was empty.
   - The problem was that the constructed SparkSql query looked something like "select from table".
   - Fixed by performing a "select count(*) from table", using the result to create an RDD of Row.empty, registering that as a temp_table_no_cols, and finally executing a "select * from temp_table_no_cols".

2. Temp tables registered with the sqlContext were never dropped.
   - The problem was that we simply never called sqlContext.dropTable()
   - Fixed by implementing a TempTable class with a .close() method, so that we can treat instances of it as managed resources (in the Scala ARM sense).

cbarca commented on loader/src/main/scala/com/actian/spark_vector/loader/command/VectorTempTable.scala:
```diff
@@ -18,7 +18,7 @@ package com.actian.spark_vector.loader.command
 
 import com.actian.spark_vector.loader.options.UserOptions
-import com.actian.spark_vector.sql.{ TableRef, VectorRelation }
+import com.actian.spark_vector.sql._
```
revert to { TableRef, VectorRelation } and add SparkSqlTable 

they're not so many individual imports to use the full ._ package instead, right?

cbarca commented on provider/src/main/scala/com/actian/spark_vector/provider/RequestHandler.scala:
```diff
@@ -126,50 +107,84 @@ class RequestHandler(sqlContext: SQLContext, val auth: ProviderAuth) extends Log
+        case "avro" => "avro"
+      }
+    }).getOrElse {
+      throw new IllegalArgumentException(s"""Could not derive format of external reference ${part.external_reference},
```
do you need the s""" ... """ ? make it s"..." if it doesn't bring nothing extra to be consistent with the rest of str formatted messages

adrian-ionescu replied:
That's pre-existing code and it's a string that's split over two lines, so I guess """ is needed after all.

cbarca replied:
hmm, ok then. my bad

cbarca commented on src/main/scala/com/actian/spark_vector/sql/SparkSqlTable.scala:
```diff
@@ -0,0 +1,30 @@
+
+sealed trait SparkSqlTable {
+  def tableName: String
+  def quotedName = sparkQuote(tableName)
```
return type

(sbt scalastyle will tell you that too for public methods, which usually get into docs)

cbarca commented on src/main/scala/com/actian/spark_vector/sql/SparkSqlTable.scala:
```diff
@@ -0,0 +1,30 @@
+}
+
+object TempTable {
+  private val id = new AtomicLong(0L)
```
\n after

cbarca commented on src/main/scala/com/actian/spark_vector/sql/SparkSqlTable.scala:
```diff
@@ -0,0 +1,30 @@
+
+object TempTable {
+  private val id = new AtomicLong(0L)
+  def apply(tableNameBase: String, df: DataFrame) = {
```
return type

(idem)

cbarca commented on provider/src/main/scala/com/actian/spark_vector/provider/RequestHandler.scala:
```diff
@@ -16,23 +16,23 @@
-import com.actian.spark_vector.loader.command.sparkQuote
-import com.actian.spark_vector.sql.VectorRelation
-import com.actian.spark_vector.util.ResourceUtil.closeResourceAfterUse
+import com.actian.spark_vector.sql._
```
if it's possible please (and not to long) import explicitly the classes you need

adrian-ionescu replied:
I think it's about 5, so most of them.. What's our policy? How many is too many?

cbarca replied:
leave it like that then, it's fine (long for me is generally something exceeding the max columns' limit a line of code, or something that is against the style adopted in that file itself)

cbarca commented on provider/src/main/scala/com/actian/spark_vector/provider/RequestHandler.scala:
```diff
@@ -66,27 +65,9 @@ class RequestHandler(sqlContext: SQLContext, val auth: ProviderAuth) extends Log
     var jobPartAccum = Future { () }
```
nit: I know it's not your code, but if `Future()` works here instead of `Future { () }` please replace it :v:

cbarca commented on src/main/scala/com/actian/spark_vector/sql/SparkSqlTable.scala:
```diff
@@ -0,0 +1,30 @@
+  override def close(): Unit = { }
+}
+
+class TempTable private (override val tableName: String, df: DataFrame) extends SparkSqlTable {
```
I don't understand the purpose of making private ctor params here and above (in HiveTable) not

and-costea replied:
For TempTables, we want to control how such objects are created (we want to be sure that the register method is called immediately after creation).

On that note, to guarantee that register isn't called multiple times, make it private.

adrian-ionescu replied:
The idea was to disallow the creation of unregistered TempTables. To be consistent, I now made register() private.

cbarca commented:
nothing more to add. ship

don't forget to:
- squash commits
- test / checkstyle
- add 'Fixes #55' near the top line of your commit (with --amend) to automatically close this issue once this pull request is merged into master (e.g. Scan no cols in provider. Fixes #55)

and-costea commented:
ship
